heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,231 @@
+  const sarifRun: ISarifRun = {
+    tool: {
+      driver: {
+        name: 'ESLint',
+        informationUri: 'https://eslint.org',
+        rules: []
+      }
+    }
+  };
+  const sarifLog: ISarifLog = {
+    version: '2.1.0',
+    $schema: 'http://json.schemastore.org/sarif-2.1.0-rtm.5',
+    runs: [sarifRun]
+  };
+
+  const { ignoreSuppressed, eslintVersion } = options;
+
+  if (typeof eslintVersion !== 'undefined') {
+    sarifRun.tool.driver.version = eslintVersion;
+  }
+
+  const sarifFiles: Map<string, ISarifFile> = new Map();
+  const sarifResults: any[] = [];
+
+  const internalErrorId = 'ESL0999';
+  const toolConfigurationNotifications: any[] = [];
+  let executionSuccessful = true;
+
+  for (const result of results) {
+    const { filePath } = result;
+    let sarifFile: ISarifFIle | undefined = sarifFiles.get(filePath);
+    if (sarifFile === undefined) {
+      const artifactIndex: number = sarifFiles.size;
+      const fileUrl: string = url.pathToFileURL(filePath).toString();
+
+      sarifFile = {
+        location: {
+          uri: fileUrl
+        }
+      };
+      sarifFiles.set(filePath, sarifFile);
+
+      const containsSuppressedMessages = result.suppressedMessages && result.suppressedMessages.length > 0;
+      const messages =
+        containsSuppressedMessages && !ignoreSuppressed
+          ? [...result.messages, ...result.suppressedMessages]
+          : result.messages;
+
+      if (messages.length > 0) {
+        for (const message of messages) {
+          const sarifRepresentation: any = {
+            level: getResultLevel(message),
+            message: {
+              text: message.message
+            },
+            locations: [
+              {
+                physicalLocation: {
+                  artifactLocation: {
+                    uri: fileUrl,
+                    index: artifactIndex
+                  }
+                }
+              }
+            ]
+          };
+
+          if (message.ruleId) {
+            sarifRepresentation.ruleId = message.ruleId;
+
+            if (containsSuppressedMessages && !ignoreSuppressed) {
+              sarifRepresentation.suppressions = message.suppressions
+                ? message.suppressions.map((suppression: Suppression) => {
+                    return {
+                      kind: suppression.kind === 'directive' ? 'inSource' : 'external',
+                      justification: suppression.justification
+                    };
+                  })
+                : [];
+            }
+          } else {
+            sarifRepresentation.descriptor = {
+              id: internalErrorId
+            };
+
+            if (sarifRepresentation.level === 'error') {
+              executionSuccessful = false;
+            }
+          }
+
+          if (message.line! > 0 || message.column! > 0) {
+            sarifRepresentation.locations[0].physicalLocation.region = {};
```

Comment:
You are indexing into this array pretty frequently. Stash the value and use it later instead of constantly reindexing